src/app/backend/handler/terminal.go

```diff
@@ -0,0 +1,264 @@
+func handleTerminalSession(session sockjs.Session) {
+  var (
+    buf string
+    err error
+    msg TerminalMessage
+    terminalSession TerminalSession
+    ok bool
+  )
+
+  if buf, err = session.Recv(); err != nil {
+    log.Printf("handleTerminalSession: can't Recv: %v", err)
+    return
+  }
+
+  if err = json.Unmarshal([]byte(buf), &msg); err != nil {
+    log.Printf("handleTerminalSession: can't UnMarshal (%v): %s", err, buf)
+    return
+  }
+
+  if msg.Op != "bind" {
+    log.Printf("handleTerminalSession: expected 'bind' message, got: %s", buf)
+    return
+  }
+
+  if terminalSession, ok = terminalSessions[msg.SessionID]; !ok {
+    log.Printf("handleTerminalSession: can't find session '%s'", msg.SessionID)
+    return
+  }
+
+  terminalSession.sockJSSession = session
+  terminalSession.bound <- nil
+  terminalSessions[msg.SessionID] = terminalSession
+}
+
+// Called from main for /api/sockjs
+func CreateAttachHandler(path string) (http.Handler) {
+  options := sockjs.DefaultOptions
+  // options.Websocket = false
+  return sockjs.NewHandler(path, options, handleTerminalSession)
+}
+
+// Called by handleAttach
+// Executed cmd in the container specified in request and connects it up with the ptyHandler (a session)
+func execHelper(request *restful.Request, cmd []string, ptyHandler PtyHandler) error {
+	namespace := request.PathParameter("namespace")
+	podName := request.PathParameter("pod")
+	containerName := request.PathParameter("container")
+
+	req := getApiClient(request).Core().RESTClient().Post().
+		Resource("pods").
+		Name(podName).
+		Namespace(namespace).
+		SubResource("exec")
+
+	req.VersionedParams(&api.PodExecOptions{
+		Container: containerName,
+		Command:   cmd,
+		Stdin:     true,
+		Stdout:    true,
+		Stderr:    true,
+		TTY:       true,
+	}, api.ParameterCodec)
+
+	exec, err := remotecommand.NewExecutor(getApiConfig(request), "POST", req.URL())
+	if err != nil {
+		return err
+	}
+
+	err = exec.Stream(remotecommand.StreamOptions{
+		SupportedProtocols: remotecommandconsts.SupportedStreamingProtocols,
+		Stdin:              ptyHandler,
+		Stdout:             ptyHandler,
+		Stderr:             ptyHandler,
+		TerminalSizeQueue:  ptyHandler,
+		Tty:                true,
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Generates a random session ID string. The format is not really interesting.
+// This ID is used to identify the session when the client opens the SockJS connection.
+// Not the same as the SockJS session id! We can't use that as that is generated
+// on the client side and we don't have it yet at this point.
+func genTerminalSessionId() (string, error) {
+	bytes := make([]byte, 16)
+	if _, err := rand.Read(bytes); err != nil {
+		return "", err
+	}
+	id := make([]byte, hex.EncodedLen(len(bytes)))
+	hex.Encode(id, bytes)
+	return string(id), nil
+}
+
+// Checks if the shell is an allowed one
+func isValidShell(validShells, shell []string) bool {
```

Comment:
Why is shell an array? Before call you do `[]string{request.QueryParameter("shell")}` which is redundant.